### PR TITLE
feat(formatting): Add max_width with array reflow

### PR DESCRIPTION
**Pull request by iepathos**
# Summary

Implements max_width support from #8. Arrays are reflowed between horizontal and vertical layouts based on width constraints, matching rustfmt behavior.

# Configuration:
- max_width: Maximum line width (default 100, matching rustfmt)
- array_width: Optional override for arrays (defaults to 60% of max_width)
- use_small_heuristics: Controls width calculation (Default/Off/Max)

# What Gets Reflowed

# Arrays:
- Collapse vertical arrays to horizontal when they fit within array_width
- Expand horizontal arrays to vertical when exceeding max_width
- Preserve comments with horizontal grouping (comments act as line-enders, matching rustfmt)
- Standalone trailing comments preserved on their own line

# What Remains Unchanged
Construct: Inline Tables
Reason: TOML 1.0 forbids newlines inside {}. Deferred until
Style Edition support (see below)

Construct: Strings
Reason: Converting to multiline alters whitespace semantics

Construct: Keys/Headers
Reason: Atomic syntax prevents splitting

Construct: Comments
Reason: Awaiting separate wrap_comments feature
Deferred Work

Inline table reflow: Requires gating on Style Edition to maintain MSRV compatibility. TOML 1.1 permits newlines in inline tables, but Cargo's TOML parser version varies by Rust edition. Will implement as follow-up with edition detection.

# Implementation

- O(n) incremental depth tracking for nesting
- unicode-width dependency for accurate character width calculation
- Gracefully handles malformed input

# Test Plan

- Vertical arrays collapse to horizontal when fitting within width
- Horizontal arrays expand to vertical when exceeding width
- Comments between elements trigger horizontal grouping
- Trailing comments on elements stay with their elements
- Standalone trailing comments preserved on own line
- Unicode combining characters handled correctly

### CODE REVIEW

**Comment by iepathos** [comment]
No problem at all. I was just trying to clean it up today before tagging you for review. It currently only has array reflow. I see that TOML 1.1 support was just merged into cargo, should we add inline table reflow to this PR as well?

**Comment by epage** [comment]
> should we add inline table reflow to this PR as well?

We'll need to add support for Style Editions and put that behind "any style edition newer than 2024" to avoid breaking people's development MSRV

Likely we should just split this out into a follow up.

File: src/lib.rs [hunk 1]
```diff
@@ -52,3 +53,50 @@ pub fn fmt_manifest(raw_input_text: &str, config: config::Config) -> Option<Stri
...
+    #[test]
+    fn already_vertical_array_stays_vertical() {
+        let input = r#"deps = [
+    "foo",
+    "bar",
+]
+"#;
+        let config = config_with_max_width(80);
+        assert_eq!(fmt_manifest(input, config).unwrap(), input);
+    }
```
Comment (epage) [inline src/lib.rs]: What was your intent with adding these tests?

File: src/formatting/overflow.rs [hunk 2]
```diff
@@ -268,4 +466,179 @@ line
         // No panic, input unchanged (can't find matching close bracket)
         assert_eq!(reflow(input, 10), input);
     }
+
+    // Additional edge cases
+
+    #[test]
+    fn dotted_key_width_included() {
```
Comment (epage) [inline src/formatting/overflow.rs]: Please add these to the previous commit

File: src/formatting/overflow.rs [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+        let expected = r#"x = [
+    [
+        [
+            1
+        ]
+    ]
+]
+"#;
+        assert_eq!(reflow(input, 5), expected);
+    }
```
Comment (epage) [inline src/formatting/overflow.rs]: nested arrays should only be reflowed if they would exceed `max_width`

File: src/formatting/overflow.rs [hunk 4]
```diff
@@ -0,0 +1,271 @@
+use crate::toml::TomlTokens;
+
+/// Reflow arrays that exceed `max_width` to vertical layout.
+#[tracing::instrument]
+pub fn reflow_arrays(_tokens: &mut TomlTokens<'_>, _max_width: usize, _tab_spaces: usize) {
```
Comment (epage) [inline src/formatting/overflow.rs]: Technically, rustfmt will
- put reflow multi-line arrays to single line
- add trailing commas to multi-line arrays; strip trailing commas from single-line arrays
- use `array_width` which by default is derived from `max_width`

Trailing commas (or not) affects the width so I'm assuming they would need to be implemented in this?  Should we do this all in one PR or do this incrementally with a PR for each incremental step?

File: src/formatting/overflow.rs (hunk 3 above)
Comment (iepathos) [reply src/formatting/overflow.rs]: Fixed, added deeply_nested_partial_reflow for test case

File: src/lib.rs (hunk 1 above)
Comment (iepathos) [reply src/lib.rs]: These were intended as integration tests to exercise fmt_manifest and the full pipeline. The original tests here focused on array reflow and two were redundant. The actual reflow behavior is covered by unit tests in overflow.rs.

I've dropped short_array_stays_horizontal and already_vertical_array_stays_vertical since they're redundant. The remaining test I've renamed to fmt_manifest_integration to clarify its intent and improved the input/expectations to check more of the pipeline (space normalization, trailing space trimming, datetime separators, blank line constraints, array reflow, and trailing commas). This boosts overall coverage ~6%.

Please let me know if you would like this setup differently or if we need better naming or comments to clarify the intent.

File: src/formatting/overflow.rs (hunk 4 above)
Comment (iepathos) [reply src/formatting/overflow.rs]: Let's handle collapse, trailing commas, and array_width in here too since they are tightly related. For mixed style arrays, we can normalize by collapsing to horizontal if it fits otherwise expanding to vertical.

File: src/lib.rs (hunk 1 above)
Comment (epage) [reply src/lib.rs]: My main interest for testing `fmt_manifest` would be
- tests specific to its functionality (e.g. skipping formatting)
- tests where multiple formatting rules have poor interactions with each other

I would skip the integration tests for this PR.

File: src/lib.rs [hunk 5]
```diff
@@ -52,3 +53,42 @@ pub fn fmt_manifest(raw_input_text: &str, config: config::Config) -> Option<Stri
...
+            "\n",
+            "\n",
+            r#"deps = ["foo", "bar", "baz", "qux"]"#,
+            "\n",
+        );
+        let expected = r#"name = "test"
+date = 2024-01-01T12:00:00
+
+deps = ["foo", "bar", "baz", "qux"]
+"#;
```
[the remark this answers is not in the document]
Comment (iepathos) [reply src/lib.rs]: Dropped these integration tests from PR

File: src/formatting/overflow.rs [hunk 6]
```diff
@@ -215,121 +801,141 @@ deps = ["foo", "bar", "baz", "qux", "quux"]
...
-        // With stub: stays horizontal
         valid(
             r#"deps = ["foo", "bar"] # comment
 "#,
             20,
             str![[r#"
-deps = ["foo", "bar"] # comment
+deps = [
+    "foo",
+    "bar"
```
Comment (epage) [inline src/formatting/overflow.rs]: Aren't we always supposed to have a trailing comma in vertical form?

File: src/formatting/overflow.rs [hunk 7]
```diff
@@ -593,38 +1236,41 @@ x = [
...
 
 "#]],
         );
     }
 
     #[test]
     fn vertical_with_comment_stays_vertical() {
+        // Don't collapse arrays with comments
         valid(
             r#"x = [
```
Comment (epage) [inline src/formatting/overflow.rs]: Unsure if I missed it but do we have any tests with mixed style arrays with comments?  The documentation talks about preserving the arrays in those cases (not just the comments)

File: src/config/mod.rs [hunk 8]
```diff
@@ -31,10 +40,14 @@ impl Default for Config {
             trailing_comma: lists::SeparatorTactic::Vertical,
             hard_tabs: false,
             tab_spaces: 4,
+            max_width: 100,
+            array_width: None,
         }
     }
 }
 
+const CONFIG_FILE_NAMES: [&str; 2] = [".rustfmt.toml", "rustfmt.toml"];
```
[the remark this answers is not in the document]
Comment (iepathos) [reply src/config/mod.rs]: Clippy was unhappy about having a const positioned after the tests I had added above it so I re-positioned it to the top. This is an unnecessary move for it, will clean this up

File: src/formatting/overflow.rs (hunk 6 above)
Comment (iepathos) [reply src/formatting/overflow.rs]: Yes, fixed

File: src/formatting/overflow.rs [hunk 9]
```diff
@@ -957,7 +1683,11 @@ x = [[1, 2, 3], [4, 5, 6]]
 "#,
             15,
             str![[r#"
-x = [[1], [2], [3]]
+x = [
+    [1],
+    [2],
+    [3]
+]
```
Comment (epage) [inline src/formatting/overflow.rs]: No trailing comma

File: src/formatting/overflow.rs [hunk 10]
```diff
@@ -0,0 +1,1091 @@
...
+"#,
+            80,
+            str![[r#"
+x = ["a", "b", # comment
+    "c",
+]
+
+"#]],
+        );
+    }
```
[the remark this answers is not in the document]
Comment (epage) [reply src/formatting/overflow.rs]: > How should we handle this?

Unless we can come up with a justification for why TOML or more specifically `Cargo.toml`, should diverge, our general aim should be to match `rustfmt`s behavior.

That doesn't mean we have to do everything in one PR though.  We can defer specific cases out to issues to be handled in follow ups, like I'm assuming we are doing with inline tables being behind a style edition.  Anything being deferred should be explicitly called out in the PR description.

[the remark this answers is not in the document]
Comment (iepathos) [reply src/formatting/overflow.rs]: Thank you, I'll add the reflow for arrays with comments in them to match rustfmt behavior

File: src/formatting/overflow.rs [hunk 11]
```diff
@@ -1,13 +1,1247 @@
...
+) -> Option<ArrayAction> {
+    // Skip arrays inside inline tables
+    if inline_table_depth > 0 {
+        return None;
+    }
+
+    let close = find_array_close(tokens, open)?;
+
+    // Skip malformed arrays (no actual closing bracket)
+    if tokens.tokens[close].raw != "]" {
```
[the remark this answers is not in the document]
Comment (iepathos) [reply src/formatting/overflow.rs]: No good reason for this inconsistency, cleaning this up

[the remark this answers is not in the document]
Comment (iepathos) [reply src/formatting/overflow.rs]: I'll improve the comment here with more detail so the issue is apparent

[the remark this answers is not in the document]
Comment (epage) [reply src/formatting/overflow.rs]: Why does the synthetic token cause a problem?
